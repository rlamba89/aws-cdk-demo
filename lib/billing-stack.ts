import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Budget } from "./constructs/budget";

interface BillingStackProps extends StackProps{
    BudgetAmount : number,
    EmailAddress : string
}

export class BillingStack extends Stack {

    constructor(scope: Construct, id: string, props: BillingStackProps){
        super(scope, id, props);

        new Budget(this, "Budget", {
            emailAddress: props.EmailAddress,
            budgetAmount:props.BudgetAmount
        })
    }
} 