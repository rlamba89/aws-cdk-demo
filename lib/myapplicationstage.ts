import { CfnOutput, Stage, StageProps } from "aws-cdk-lib"
import { Construct } from "constructs"
import { ServerStack } from "./server-stack"

interface IApplicationStageProps extends StageProps{
    StageName:string
}

export class MyApplicationStage extends Stage{
    public readonly urlOutput: CfnOutput;
    constructor(scope: Construct, id: string, props: IApplicationStageProps) {
        super(scope, id, props)
        
        const lambdaServerStack = new ServerStack(this, `CDK-Lambda-Stack-${props.StageName}`,{
            StageName : props.StageName
        })

        this.urlOutput = lambdaServerStack.urlOutput
    }
}