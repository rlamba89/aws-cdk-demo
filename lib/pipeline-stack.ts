import { SecretValue, Stack, StackProps } from 'aws-cdk-lib';
import { BuildSpec, LinuxBuildImage, PipelineProject } from 'aws-cdk-lib/aws-codebuild';
import {  Pipeline, Artifact } from "aws-cdk-lib/aws-codepipeline";
import { CloudFormationCreateUpdateStackAction, CodeBuildAction, GitHubSourceAction } from 'aws-cdk-lib/aws-codepipeline-actions';
import { Construct } from 'constructs';


export class PipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const pipeline = new Pipeline(this, 'Pipeline',{
      pipelineName:'CDK-Demo-Pipeline',
      crossAccountKeys:false
    });

    const cdkPipelineSourceOutput =  new Artifact('CDK-Demo-Pipeline-Source-Output')
    const cdkServiceSourceOutput = new Artifact('CDK-Demo-Service-Source-Output')

    pipeline.addStage({
      stageName: "Source",
      actions: [
        new GitHubSourceAction({
          owner: "rlamba89",
          repo: "aws-cdk-demo",
          branch: "master",
          actionName: "CDK-Demo-Pipeline-Source",
          oauthToken: SecretValue.secretsManager("aws_github_cdk_demo"),
          output: cdkPipelineSourceOutput,
        }),
        new GitHubSourceAction({
          owner: "rlamba89",
          repo: "express-lambda",
          branch: "master",
          actionName: "CDK-Demo-Service-Source",
          oauthToken: SecretValue.secretsManager("aws_github_cdk_demo"),
          output: cdkServiceSourceOutput,
        })
      ],
    });

    const cdkBuildOutput = new Artifact('CDK-Demo-Build-Output')
    const cdkServiceBuildOutput = new Artifact('CDK-Demo-Service-Build-Output')


    pipeline.addStage({
      stageName: 'Build',
      actions:[
        new CodeBuildAction({
          actionName:'CDK-Build',
          input:cdkPipelineSourceOutput,
          outputs:[cdkBuildOutput],
          project:new PipelineProject(this, 'CdkBuildProject',{
            environment:{
              buildImage:LinuxBuildImage.STANDARD_5_0
            },
            buildSpec:BuildSpec.fromSourceFilename('build-specs/cdk-build-spec.yml')
          })
        }),
        new CodeBuildAction({
          actionName:'CDK-Service-Build',
          input:cdkServiceSourceOutput,
          outputs:[cdkServiceBuildOutput],
          project:new PipelineProject(this, 'CdkServiceBuildProject',{
            environment:{
              buildImage:LinuxBuildImage.STANDARD_5_0
            },
            buildSpec:BuildSpec.fromSourceFilename('build-specs/cdk-service-build-spec-yml')
          })
        })

      ]
    })

    pipeline.addStage({
      stageName:'Pipeline_Update',
      actions:[
        new CloudFormationCreateUpdateStackAction({
          actionName:'CDK-Pipeline-Update',
          stackName:'PipelineStack',
          templatePath:cdkBuildOutput.atPath('PipelineStack.template.json'),
          adminPermissions:true
        })
      ]
    })
  }
}
