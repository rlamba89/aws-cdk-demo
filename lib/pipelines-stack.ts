import { pipelines, SecretValue, Stack, StackProps, Stage, StageProps } from "aws-cdk-lib";
import { Artifact } from "aws-cdk-lib/aws-codepipeline";
import { GitHubTrigger } from "aws-cdk-lib/aws-codepipeline-actions";
import { ManualApprovalStep } from "aws-cdk-lib/pipelines";
import { Construct } from "constructs";
import { MyApplicationStage } from "./myapplicationstage";
import { ServerStack } from "./server-stack";


export class PipelinesStack extends Stack{
    constructor(scope:Construct, id: string, props?:StackProps){
        super(scope, id, props)

        const cdkPipelineSourceOutput =  new Artifact('CDK-Demo-Pipeline-Source-Output')

        const pipeline = new pipelines.CodePipeline(this, 'Pipeline-v2', {
                pipelineName:'CDK-Demo-Pipeline-v2',
                selfMutation:true,
            
                synth: new pipelines.ShellStep('Synth', {
                    input : pipelines.CodePipelineSource.gitHub("rlamba89/aws-cdk-demo","master",{
                        authentication:SecretValue.secretsManager("aws_github_cdk_demo"),
                        trigger:GitHubTrigger.WEBHOOK
                    }),
                    commands: [
                      'npm install -g npm',
                      'npm install',
                      'npm run build',
                      'npm test',
                      'npm run cdk -- synth',
                    ],
                  }),
        });
        //new pipelines.p
        
        const UAT = pipeline.addStage(new MyApplicationStage(this, "Test",{
            StageName:'Test'
        }));

         UAT.addPost(new ManualApprovalStep ("Approve release to prod"));

        const Prod = pipeline.addStage(new MyApplicationStage(this, "Prod",{
            StageName : 'Prod'
        }));

    }
}

