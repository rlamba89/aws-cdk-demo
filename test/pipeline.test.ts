import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { PipelineStack } from "../lib/pipeline-stack";

test('Pipeline test', () =>{
    const app = new App();

    //when
    const stack = new PipelineStack(app, 'MyTestPipelineStack')
    
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
});