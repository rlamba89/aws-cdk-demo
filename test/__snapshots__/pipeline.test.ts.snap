// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline test 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "CdkBuildProject9382C38D": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "CODEPIPELINE",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "CdkBuildProjectRoleE0B6FEB0",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "build-specs/cdk-build-spec.yml",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "CdkBuildProjectRoleDefaultPolicy3C7ECB00": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "CdkBuildProject9382C38D",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "CdkBuildProject9382C38D",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "CdkBuildProject9382C38D",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkBuildProjectRoleDefaultPolicy3C7ECB00",
        "Roles": Array [
          Object {
            "Ref": "CdkBuildProjectRoleE0B6FEB0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkBuildProjectRoleE0B6FEB0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaBuildProject7E2DAB11": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "CODEPIPELINE",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "LambdaBuildProjectRoleD0C4F982",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "build-specs/lambda-build-spec.yml",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "LambdaBuildProjectRoleD0C4F982": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaBuildProjectRoleDefaultPolicyA3A66624": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "LambdaBuildProject7E2DAB11",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "LambdaBuildProject7E2DAB11",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "LambdaBuildProject7E2DAB11",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaBuildProjectRoleDefaultPolicyA3A66624",
        "Roles": Array [
          Object {
            "Ref": "LambdaBuildProjectRoleD0C4F982",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineArtifactsBucket22248F97": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "PipelineArtifactsBucketPolicyD4F9712A": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "PipelineArtifactsBucket22248F97",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "PipelineBuildCDKBuildCodePipelineActionRoleDefaultPolicy0BFA6CD6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "CdkBuildProject9382C38D",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineBuildCDKBuildCodePipelineActionRoleDefaultPolicy0BFA6CD6",
        "Roles": Array [
          Object {
            "Ref": "PipelineBuildCDKBuildCodePipelineActionRoleF76A3E0C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineBuildCDKBuildCodePipelineActionRoleF76A3E0C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineBuildLambdaBuildCodePipelineActionRole94BC7353": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineBuildLambdaBuildCodePipelineActionRoleDefaultPolicy733B90FF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "LambdaBuildProject7E2DAB11",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineBuildLambdaBuildCodePipelineActionRoleDefaultPolicy733B90FF",
        "Roles": Array [
          Object {
            "Ref": "PipelineBuildLambdaBuildCodePipelineActionRole94BC7353",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineC660917D": Object {
      "DependsOn": Array [
        "PipelineRoleDefaultPolicyC7A05455",
        "PipelineRoleD68726F7",
      ],
      "Properties": Object {
        "ArtifactStore": Object {
          "Location": Object {
            "Ref": "PipelineArtifactsBucket22248F97",
          },
          "Type": "S3",
        },
        "Name": "CDK-Demo-Pipeline",
        "RestartExecutionOnUpdate": true,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "PipelineRoleD68726F7",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "ThirdParty",
                  "Provider": "GitHub",
                  "Version": "1",
                },
                "Configuration": Object {
                  "Branch": "master",
                  "OAuthToken": "{{resolve:secretsmanager:aws_github_cdk_demo:SecretString:::}}",
                  "Owner": "rlamba89",
                  "PollForSourceChanges": false,
                  "Repo": "aws-cdk-demo",
                },
                "Name": "CDK-Demo-Pipeline-Source",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "CDK-Demo-Pipeline-Source-Output",
                  },
                ],
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "CdkBuildProject9382C38D",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "CDK-Demo-Pipeline-Source-Output",
                  },
                ],
                "Name": "CDK-Build",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "CDK-Demo-Build-Output",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "PipelineBuildCDKBuildCodePipelineActionRoleF76A3E0C",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "LambdaBuildProject7E2DAB11",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "CDK-Demo-Pipeline-Source-Output",
                  },
                ],
                "Name": "Lambda-Build",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "CDK-Demo-Service-Build-Output",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "PipelineBuildLambdaBuildCodePipelineActionRole94BC7353",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CloudFormation",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ActionMode": "CREATE_UPDATE",
                  "Capabilities": "CAPABILITY_NAMED_IAM",
                  "RoleArn": Object {
                    "Fn::GetAtt": Array [
                      "PipelinePipelineUpdatePipelineUpdateRoleC024FEB0",
                      "Arn",
                    ],
                  },
                  "StackName": "PipelineStack",
                  "TemplatePath": "CDK-Demo-Build-Output::PipelineStack.template.json",
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "CDK-Demo-Build-Output",
                  },
                ],
                "Name": "Pipeline-Update",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole55089D01",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Pipeline_Update",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole55089D01": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRoleDefaultPolicy83C64A65": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelinePipelineUpdatePipelineUpdateRoleC024FEB0",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "cloudformation:CreateStack",
                "cloudformation:DescribeStack*",
                "cloudformation:GetStackPolicy",
                "cloudformation:GetTemplate*",
                "cloudformation:SetStackPolicy",
                "cloudformation:UpdateStack",
                "cloudformation:ValidateTemplate",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":cloudformation:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":stack/PipelineStack/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRoleDefaultPolicy83C64A65",
        "Roles": Array [
          Object {
            "Ref": "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole55089D01",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelinePipelineUpdatePipelineUpdateRoleC024FEB0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudformation.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelinePipelineUpdatePipelineUpdateRoleDefaultPolicy1857F0DC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelinePipelineUpdatePipelineUpdateRoleDefaultPolicy1857F0DC",
        "Roles": Array [
          Object {
            "Ref": "PipelinePipelineUpdatePipelineUpdateRoleC024FEB0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineRoleD68726F7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineRoleDefaultPolicyC7A05455": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineBuildCDKBuildCodePipelineActionRoleF76A3E0C",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineBuildLambdaBuildCodePipelineActionRole94BC7353",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelinePipelineUpdatePipelineUpdateCodePipelineActionRole55089D01",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineRoleDefaultPolicyC7A05455",
        "Roles": Array [
          Object {
            "Ref": "PipelineRoleD68726F7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineSourceCDKDemoPipelineSourceWebhookResourceD8F9DB0C": Object {
      "Properties": Object {
        "Authentication": "GITHUB_HMAC",
        "AuthenticationConfiguration": Object {
          "SecretToken": "{{resolve:secretsmanager:aws_github_cdk_demo:SecretString:::}}",
        },
        "Filters": Array [
          Object {
            "JsonPath": "$.ref",
            "MatchEquals": "refs/heads/{Branch}",
          },
        ],
        "RegisterWithThirdParty": true,
        "TargetAction": "CDK-Demo-Pipeline-Source",
        "TargetPipeline": Object {
          "Ref": "PipelineC660917D",
        },
        "TargetPipelineVersion": 1,
      },
      "Type": "AWS::CodePipeline::Webhook",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
